### PR TITLE
Typescript manifest

### LINKED ISSUE #1938: Support TS based project schemas
### Description
As the project manifest gets more complex its harder to know what options are needed.

Users could define their project manifests in typescript files with typescript types and rely on IDE support for typescript
An article explaining another projects decisions: https://stellate.co/blog/moving-from-yaml-to-typescript

### Advantages
* Probably more flexible/powerful type/schema control than JSON-Schema

### Disadvantages
* Tied to SDK support and compatibility
* Possible security issues with loading untrusted code and leaking sensitive data

## Requirements

* Create a `types-core` package to pull interfaces/types from common. As well as the store interface. types-* will use this as a dependency
* Add JSDOC descriptions and examples to fields
* Update publish command to transform TS to yaml deployment
* Update loading to support TS manifests

### PR DESCRIPTION
**Pull request by jiqiang90**
# Description
Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes #1938 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist

- [ ] I have tested locally
- [ ] I have performed a self review of my changes
- [ ] Updated any relevant documentation
- [ ] Linked to any relevant issues
- [ ] I have added tests relevant to my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] My code is up to date with the base branch
- [ ] I have updated relevant changelogs. [We suggest using chan](https://github.com/geut/chan/tree/main/packages/chan)

### CODE REVIEW

File: packages/types-core/src/store.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+export enum OperationType {
+  Set = 'Set',
+  Remove = 'Remove',
+}
+
+export type OperationEntity = {
+  operation: OperationType;
+  entityType: string;
+  data: Entity | string;
+};
```
Comment (stwiname) [inline packages/types-core/src/store.ts:13]: I think these can remain inside node core. They aren't used externally.

**Review by stwiname — COMMENTED** [review]
Nice work

* I think it would be good to update codegen to use `Entity` and `FunctionPropertyNames` from '@subql/types-core`
* I think we can remove the versioned manifests below 1.0.0. They aren't supported now

File: packages/types-core/src/project/types.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+export interface CommonSubqueryNetworkConfig extends Omit<IProjectNetworkConfig, 'endpoint'> {
+  network?: string | string[];
+  chainTypes?: FileReference;
+}
+
+// Expose to user manifest
+export interface CommonSubqueryProject<
+  N extends CommonSubqueryNetworkConfig = CommonSubqueryNetworkConfig,
+  DS extends BaseDataSource = BaseDataSource,
+  T = unknown
```
[the remark this answers is not in the document]
Comment (jiqiang90) [reply packages/types-core/src/project/types.ts:34]: I think it was a mistake, does any network object have a name?

[the remark this answers is not in the document]
Comment (stwiname) [reply packages/types-core/src/project/types.ts:34]: Yes. The name is used to identify the template

File: package.json [hunk 3]
```diff
@@ -4,6 +4,9 @@
   "workspaces": [
     "packages/*"
   ],
+  "dependencies": {
+    "typescript": "^4.9.5"
```
Comment (stwiname) [inline package.json:8]: Should this be under the package that needs the sandbox?

File: packages/node-core/test/sandbox/tsconfig.json [hunk 4]
```diff
@@ -16,5 +16,5 @@
     "outDir": "."
   },
   "exclude": [],
-  "include": ["./main.ts"]
+  "include": ["./main.ts", "./project.ts"]
```
Comment (stwiname) [inline packages/node-core/test/sandbox/tsconfig.json:19]: Will users have to update their ts config if they use a ts manifest?

File: packages/types-core/CHANGELOG.md [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+All notable changes to this project will be documented in this file.
+
+The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
+and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
+
+## [Unreleased]
+
+[Unreleased]: https://github.com/subquery/subql/compare/types/2.2.0...HEAD
+[2.2.0]: https://github.com/subquery/subql/compare/types/2.1.4...types/2.2.0
+[0.2.0]: https://github.com/subquery/subql/tag/v0.2.0
```
Comment (stwiname) [inline packages/types-core/CHANGELOG.md:11]: I think these can go

File: packages/types-core/package.json [hunk 6]
```diff
@@ -0,0 +1,22 @@
+{
+  "name": "@subql/types-core",
+  "version": "2.2.1-0",
```
Comment (stwiname) [inline packages/types-core/package.json:3]:
```suggestion
  "version": "0.0.0-0",
```
Then we won't start on some random version

File: packages/types-core/src/project/versioned/index.ts [hunk 7]
```diff
@@ -0,0 +1,7 @@
+// Copyright 2020-2023 SubQuery Pte Ltd authors & contributors
+// SPDX-License-Identifier: GPL-3.0
+
+export * from './base';
+export * from './v0_2_0';
+export * from './v0_2_1';
```
Comment (stwiname) [inline packages/types-core/src/project/versioned/index.ts:6]: I thought we were removing these?

File: packages/types-core/test/project.ts [hunk 8]
```diff
@@ -0,0 +1,40 @@
+// Copyright 2020-2023 SubQuery Pte Ltd authors & contributors
+// SPDX-License-Identifier: GPL-3.0
+
+// import {ProjectManifestV1_0_0} from '@subql/common';
+// import {ProjectManifestV1_0_0Impl} from '@subql/common-substrate';
+// import {ProjectManifestImpls} from './ProjectManifestVersioned'
+// import {SubstrateDatasource} from '@subql/types';
+
+//
+// const manifest: ProjectManifestV1_0_0<SubstrateDatasource> = {
```
Comment (stwiname) [inline packages/types-core/test/project.ts:10]: Should users be using a specific version here? I thought it was `SubstrateProject`?

**Comment by stwiname** [comment]
Doing some testing:
* `build`, `codegen` and `publish` don't work if you use a directory with just a ts file and no yaml
* `publish` fails with a path to ts file. `Error: Unable to parse project manifest`

File: packages/node-core/test/sandbox/tsconfig.json (hunk 4 above)
Comment (jiqiang90) [reply packages/node-core/test/sandbox/tsconfig.json:19]: No, it doesn't require user project add this to tsconfig. 

This line was added to build `project.js`, in order for testing purpose.

File: packages/types-core/test/project.ts (hunk 8 above)
Comment (jiqiang90) [reply packages/types-core/test/project.ts:10]: Was a draft, this can be removed now

**Comment by jiqiang90** [comment]
> Doing some testing:
> 
> * `build`, `codegen` and `publish` don't work if you use a directory with just a ts file and no yaml
Now default manifest is in yaml, maybe we can add logic to look ts first, if not found then try yaml. 

> * `publish` fails with a path to ts file. `Error: Unable to parse project manifest`

image

File: packages/cli/src/commands/publish.ts [hunk 9]
```diff
@@ -18,22 +18,43 @@ export default class Publish extends Command {
   static description = 'Upload this SubQuery project to IPFS';
 
   static flags = {
-    location: Flags.string({char: 'f', description: 'from project or manifest path'}),
+    file: Flags.string({
+      char: 'f',
+      description: 'from specify manifest file path (will overwrite -l if both used)',
+    }),
+    location: Flags.string({char: 'l', description: 'from project folder'}),
```
Comment (stwiname) [inline packages/cli/src/commands/publish.ts:25]: I think we should change this back

File: packages/cli/src/commands/publish.ts [hunk 10]
```diff
@@ -18,22 +18,43 @@ export default class Publish extends Command {
...
+        await Build.run(['--file', flags.file, '-s']);
+        hasBuilt = true;
+      }
+    }
+    // This can be a file path or dir path
+    const location = flags.file
+      ? resolveToAbsolutePath(flags.file)
+      : flags.location
+      ? resolveToAbsolutePath(flags.location)
+      : process.cwd();
```
Comment (stwiname) [inline packages/cli/src/commands/publish.ts:47]: We have the same logic in build

File: packages/cli/src/commands/publish.ts [hunk 11]
```diff
@@ -18,22 +18,43 @@ export default class Publish extends Command {
...
-    } catch (e) {
-      this.log(e);
-      this.error('Failed to build project');
+    if (!hasBuilt) {
+      try {
+        await Build.run(['--location', project.root, '-s']);
+      } catch (e) {
+        this.log(e);
+        this.error('Failed to build project');
+      }
```
Comment (stwiname) [inline packages/cli/src/commands/publish.ts:57]: I think we can just have a single build call

File: packages/common/src/project/utils.ts [hunk 12]
```diff
@@ -254,3 +268,6 @@ export class FileReferenceImp<T> implements ValidatorConstraintInterface {
     return typeof fileReference === 'object' && 'file' in fileReference && typeof fileReference.file === 'string';
   }
 }
+
+export const tsProjectYamlPath = (tsManifestEntry: string) =>
+  path.join(path.dirname(tsManifestEntry), `${path.basename(tsManifestEntry, path.extname(tsManifestEntry))}.yaml`);
```
Comment (stwiname) [inline packages/common/src/project/utils.ts:273]: Can this be 
```ts
tsManifestEntry.replace('.ts', '.yaml');
```

File: packages/types-core/src/project/versioned/v1_0_0/types.ts [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+
+/**
+ * Represents optional configuration options for a node.
+ * @interface
+ */
+export interface NodeOptions {
+  /**
+   * Indicates whether historical state tracking is enabled for the indexer (optional).
+   * @type {boolean}
+   * @default "true"
```
Comment (jamesbayly) [inline packages/types-core/src/project/versioned/v1_0_0/types.ts:85]: Is this default correct? I thought we defaulted to enabling historical?

File: packages/types-core/src/project/versioned/v1_0_0/types.ts [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+   * @type {boolean}
+   * @default "false"
+   */
+  unfinalizedBlocks?: boolean;
+
+  /**
+   * Indicates whether block skipping is enabled for the indexer (optional).
+   * @type {boolean}
+   */
+  skipBlock?: boolean;
```
Comment (jamesbayly) [inline packages/types-core/src/project/versioned/v1_0_0/types.ts:114]: There's no docs for this command anywhere?

File: packages/types/src/project.ts [hunk 15]
```diff
@@ -22,108 +60,276 @@ export type RuntimeHandlerInputMap<T extends AnyTuple = AnyTuple> = {
...
 }
 
+/**
+ * Represents a filter for Substrate events, extending SubstrateBaseHandlerFilter.
+ * @interface
+ * @extends {SubstrateBaseHandlerFilter}
+ */
 export interface SubstrateEventFilter extends SubstrateBaseHandlerFilter {
+  /**
+   * The module name for filtering events or calls (optional).
```
Comment (jamesbayly) [inline packages/types/src/project.ts:105]: Is this cases sensitive?

File: packages/types/src/project.ts [hunk 16]
```diff
@@ -22,108 +60,276 @@ export type RuntimeHandlerInputMap<T extends AnyTuple = AnyTuple> = {
...
+   * @type {string}
+   * @example
+   * module: 'balances'
+   */
   module?: string;
+
+  /**
+   * The method name for filtering events calls (optional).
```
Comment (jamesbayly) [inline packages/types/src/project.ts:113]: Is this case sensitive? If so
```suggestion
   * The method name for filtering events calls (case-insensitive) (optional).
```

File: packages/types-core/src/project/versioned/v1_0_0/types.ts (hunk 14 above)
Comment (stwiname) [reply packages/types-core/src/project/versioned/v1_0_0/types.ts:114]: https://github.com/subquery/documentation/pull/400

File: packages/types-core/src/project/versioned/v1_0_0/types.ts (hunk 13 above)
Comment (stwiname) [reply packages/types-core/src/project/versioned/v1_0_0/types.ts:85]: Yes, historical is enabled by default

File: packages/types-core/src/project/versioned/v1_0_0/types.ts [hunk 17]
```diff
@@ -0,0 +1,167 @@
...
+   */
+  historical?: boolean;
+
+  /**
+   * Indicates whether unsafe mode are allowed for the indexer (optional).
+   * 
+   * Unsafe mode controls various features that compromise the determinsm of a SubQuery project.
+   * Unsafe allows any imports which greatly increases functionality with the tradeoff of decreased security
+   * @type {boolean}
+   * @default "false"
```
[the remark this answers is not in the document]
Comment (stwiname) [reply packages/types-core/src/project/versioned/v1_0_0/types.ts:95]: Unsafe and unfinalized blocks are false by default

File: packages/types/src/project.ts (hunk 15 above)
Comment (stwiname) [reply packages/types/src/project.ts:105]: Module and method are both case sensitive

Comment (jiqiang90) [reply packages/types/src/project.ts:105]: see https://github.com/subquery/subql/pull/1706

File: packages/node-core/src/indexer/blockDispatcher/base-block-dispatcher.ts [hunk 18]
```diff
@@ -5,7 +5,17 @@ import assert from 'assert';
...
+  DynamicDsService,
+  IProjectService,
+  PoiBlock,
+  PoiService,
+  StoreCacheService,
+  SmartBatchService,
+  StoreService,
+} from '..';
+import {NodeConfig} from '../../configure';
+import {IProjectUpgradeService, NodeConfig} from '../../configure/ProjectUpgrade.service';
```
Comment (stwiname) [inline packages/node-core/src/indexer/blockDispatcher/base-block-dispatcher.ts:18]: This is causing build issues

**Review by stwiname — APPROVED** [review]

**Pull request merged**